### PR TITLE
[SPARK-34639][SQL] Always remove unnecessary Alias in Analyzer.resolveExpression

### PR DESCRIPTION
**Pull request by cloud-fan**
### What changes were proposed in this pull request?

In `Analyzer.resolveExpression`, we have a parameter to decide if we should remove unnecessary `Alias` or not. This is over complicated and we can always remove unnecessary `Alias`.

This PR simplifies this part and removes the parameter.

### Why are the changes needed?

code cleanup

### Does this PR introduce _any_ user-facing change?

no

### How was this patch tested?

existing tests

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 1]
```diff
@@ -1964,28 +1963,22 @@ class Analyzer(override val catalogManager: CatalogManager)
...
       e,
       q,
       resolveColumnByName = nameParts => {
         q.resolveChildren(nameParts, resolver)
       },
       resolveColumnByOrdinal = ordinal => {
-        val candidates = q.children.flatMap(_.output)
-        assert(ordinal >= 0 && ordinal < candidates.length)
-        candidates.apply(ordinal)
+        assert(q.children.length == 1)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:1978]: This change is not related to this PR, but a small followup from https://github.com/apache/spark/pull/31728#discussion_r588049629

**Comment by cloud-fan** [comment]
cc @maropu @viirya

**Review by dongjoon-hyun — COMMENTED** [review]
Could you check the UT failures?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 2]
```diff
@@ -1905,11 +1904,13 @@ class Analyzer(override val catalogManager: CatalogManager)
...
-            // CleanupAliases later in Analyzer, trim non top-level unnecessary alias of
-            // `GetStructField` here is safe.
-            case Alias(s: GetStructField, _) if trimAlias && !isTopLevel => s
+            // We trim unnecessary alias of `Get[Array]StructField` here. Note that, we cannot trim
+            // the alias of top-level `Get[Array]StructField`, as we should resolve
+            // `UnresolvedAttribute` to a named expression. The caller side can trim the alias of
+            // top-level `GetStructField` if it's safe to do so. Since we will call CleanupAliases
+            // later in Analyzer, trim non top-level unnecessary alias here is safe.
+            case Alias(s: GetStructField, _) if !isTopLevel => s
+            case Alias(s: GetArrayStructFields, _) if !isTopLevel => s
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:1913]: It looks we also need to add entries for the other `ExtractValue` classes, e.g., `GetMapValue`?
They seems to have the same issue with SPARK-31670;
```
scala> spark.table("t").printSchema()
root
 |-- c0: integer (nullable = false)
 |-- c1: map (nullable = false)
 |    |-- key: string
 |    |-- value: string (valueContainsNull = true)


scala> sql("select c0, c1.key, COUNT(1) from t group by c0, c1.key with cube").show()
org.apache.spark.sql.AnalysisException: expression 't.`c1`' is neither present in the group by, nor is it an aggregate function. Add to group by or wrap in first() (or first_value) if you don't care which value you get.;
Aggregate [c0#34, key#35, spark_grouping_id#33L], [c0#34, c1#24[key] AS key#28, count(1) AS count(1)#30L]
+- Expand [List(c0#23, c1#24, c0#31, key#32, 0), List(c0#23, c1#24, c0#31, null, 1), List(c0#23, c1#24, null, key#32, 2), List(c0#23, c1#24, null, null, 3)], [c0#23, c1#24, c0#34, key#35, spark_grouping_id#33L]
   +- Project [c0#23, c1#24, c0#23 AS c0#31, c1#24[key] AS key#32]
      +- SubqueryAlias t
         +- View (`t`, [c0#23,c1#24])
            +- Project [cast(col1#25 as int) AS c0#23, cast(col2#26 as map<string,string>) AS c1#24]
               +- Project [col1#25, col2#26]
                  +- LocalRelation [col1#25, col2#26]
```

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:1913]: good catch!

File: sql/core/src/main/scala/org/apache/spark/sql/RelationalGroupedDataset.scala [hunk 3]
```diff
@@ -80,11 +80,7 @@ class RelationalGroupedDataset protected[sql](
     }
   }
 
-  // Wrap UnresolvedAttribute with UnresolvedAlias, as when we resolve UnresolvedAttribute, we
-  // will remove intermediate Alias for ExtractValue chain, and we need to alias it again to
-  // make it a NamedExpression.
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/RelationalGroupedDataset.scala:85]: The comment is wrong as we don't remove top-level aliases for aggregate expressions. It causes problems as it wraps `UnresolvedAttribute` with `UnresolvedAlias`, making it not top-level anymore. Then the alias will be removed after this patch and `UnresolvedAlias` generates a different name.

For nested field `a.b`, previously the resolved expression is `Alias(GetStructField(...), "b")` and the `Alias` is not removed. `UnresolvedAlias` is useless and will be simply removed. So the final output column name is `b`. Now we remove the `Alias`, and `UnresolvedAlias` kicks in and generates a new `Alias` with the name `a.b`, which is a behavior change.

Here I simply remove this `UnresolvedAlias`, to make the behavior the same as before.

**Comment by cloud-fan** [comment]
retest this please

**Comment by cloud-fan** [comment]
any more comments? @viirya @maropu

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 4]
```diff
@@ -3975,8 +3964,15 @@ object ResolveCreateNamedStruct extends Rule[LogicalPlan] {
   override def apply(plan: LogicalPlan): LogicalPlan = plan.resolveExpressions {
     case e: CreateNamedStruct if !e.resolved =>
       val children = e.children.grouped(2).flatMap {
-        case Seq(NamePlaceholder, e: NamedExpression) if e.resolved =>
-          Seq(Literal(e.name), e)
+        case Seq(NamePlaceholder, ne: NamedExpression) if ne.resolved =>
+          Seq(Literal(ne.name), ne)
+        case Seq(NamePlaceholder, g @ GetStructField(child, _, Some(name))) if child.resolved =>
```
[the remark this answers is not in the document]
Comment (dongjoon-hyun) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:3969]: Could you add a test case for this code path?

[the remark this answers is not in the document]
Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:3969]: Only `GetStructField`, `GetArrayStructFields` and `GetMapValue` are the possible expressions here?

Can it be `case Seq(NamePlaceholder, other expression)`?

[the remark this answers is not in the document]
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:3969]: then let's have a separated PR for it: https://github.com/apache/spark/pull/31808

File: sql/core/src/main/scala/org/apache/spark/sql/RelationalGroupedDataset.scala (hunk 3 above)
Comment (viirya) [reply sql/core/src/main/scala/org/apache/spark/sql/RelationalGroupedDataset.scala:85]: Hmm, this code is pretty old, 2015.

I saw `alias` is also used to add alias around grouping expressions, not just aggregate expressions. Seems the comment is more for the case?

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/RelationalGroupedDataset.scala:85]: This is only for `Aggregate.aggregateExpressions` not `Aggregate.groupingExpressions`. `Aggregate.aggregateExpressions` can include grouping expressions, but it doesn't matter. It needs to be `Seq[NamedExpression]` and Spark won't remove the top-level alias in it.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 4 above)
[the remark this answers is not in the document]
Comment (dongjoon-hyun) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:3969]: Thank you for splitting.

**Comment by dongjoon-hyun** [comment]
Could you resolve the conflicts?

**Comment by maropu** [comment]
I've checked that code and the change itself looks fine.

**Review by maropu — APPROVED** [review]

**Review by viirya — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks for the review, merging to master!

**Pull request closed without merging**